### ISSUE TITLE
Converting torch module with non-constant parameters

### ISSUE THREAD

**Issue by shivadb** [NVIDIA-AI-IOT/torch2trt#584]
Here is a simple toy model that simply stores the :

         class InputCache(nn.Module):
             def __init__(self, tensor):
                 super(InputCache, self).__init__()
                 self.register_buffer('cache', tensor)
             
             def forward(self, x):
                 cache_update = torch.cat((self.cache[:,:,1:], x.detach()), dim=2)
                 self.cache = cache_update
                 return self.cache

        tc = InputCache(torch.zeros(1,1,10))

The above example simply stores the last 10 input values. 
If I try to convert tc to tensort using `tc_trt = torch2trt(tc, torch.tensor([[[0.]]]).cuda())`, it will give dimension errors (I think torch2trt removes the batch dimension of the input). I can work around this by reducing the dimension of the cache. But I then get an attribute error: 'Tensor' object has no attribute '_trt'

So my understanding is that I need to create a custom converter for this module. However, I cannot find any examples where the module's state (parameters, buffers) change during runtime. How would I go about converting this type of model?

**Comment by jaybdub** [comment]
Hi shivadb,

Thanks for reaching out!

Unfortunately, TensorRT doesn't support stateful engines (that change behavior from input to input).

However, you can get around this by pulling the caching logic outside of the engine, and providing the cached data as inputs to the TensorRT engine.

Are you able to share how you're intending to use this module in your larger pipeline?  This might help indicate the best method for moving forward.

Best,
John

**Comment by shivadb** [comment]
Thanks for the feedback John. I was trying to include the caching logic within the engine because I was trying to implement such a module at several points the network and was trying to avoid separating the network at each spot this layer is implemented. 

I thought this might be possible because I was reading that TensorRT has a RNN layer ([here](https://docs.nvidia.com/deeplearning/tensorrt/developer-guide/index.html#rnnv2-layer)) which I assumed was stateful.

I will go ahead and try to implement this logic outside the engine.

Thanks,
Shiv
